sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -194,47 +271,65 @@ class Response
 {
 
 private:
-  uint16_t m_statusCode;
+  uint16_t m_mayorVersion;
+  uint16_t m_minorVersion;
+  HttpStatusCode m_statusCode;
   std::string m_reasonPhrase;
   std::map<std::string, std::string> m_headers;
 
   // Response can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
-  http::BodyBuffer* m_bodyBuffer;
+  std::vector<uint8_t> m_bodyBuffer;
   http::BodyStream* m_bodyStream;
 
   Response(
-      uint16_t statusCode,
+      uint16_t mayorVersion,
+      uint16_t minorVersion,
+      HttpStatusCode statusCode,
       std::string const& reasonPhrase,
-      BodyBuffer* const bodyBuffer,
+      std::vector<uint8_t> const& bodyBuffer,
       BodyStream* const BodyStream)
-      : m_statusCode(statusCode), m_reasonPhrase(reasonPhrase), m_bodyBuffer(bodyBuffer),
-        m_bodyStream(BodyStream)
+      : m_mayorVersion(mayorVersion), m_minorVersion(minorVersion), m_statusCode(statusCode),
+        m_reasonPhrase(reasonPhrase), m_bodyBuffer(bodyBuffer), m_bodyStream(BodyStream)
   {
   }
 
 public:
-  Response(uint16_t statusCode, std::string const& reasonPhrase)
-      : Response(statusCode, reasonPhrase, http::BodyBuffer::null, http::BodyStream::null)
+  Response(
+      uint16_t mayorVersion,
+      uint16_t minorVersion,
+      HttpStatusCode statusCode,
+      std::string const& reasonPhrase)
+      : Response(
+            mayorVersion,
+            minorVersion,
+            statusCode,
+            reasonPhrase,
+            std::vector<uint8_t>(),
+            http::BodyStream::null)
   {
   }
 
-  // Methods used to build HTTP response
-  void addHeader(std::string const& name, std::string const& value);
-  void setBody(BodyBuffer* bodyBuffer);
-  void setBody(BodyStream* bodyStream);
-
-  // Methods used by transport layer (and logger) to send response
-  uint16_t getStatusCode();
-  std::string const& getReasonPhrase();
-  std::map<std::string, std::string> const& getHeaders();
-  http::BodyStream* getBodyStream();
-  http::BodyBuffer* getBodyBuffer();
+  // Methods used by transport layer upon receiving a response
+  void AddHeader(std::string const& name, std::string const& value);
+  void AppendBody(uint8_t* ptr, uint64_t size);
+
+  // Util methods for customers to read response/parse an http response
+  HttpStatusCode GetStatusCode();
+  std::string const& GetReasonPhrase();
+  std::map<std::string, std::string> const& GetHeaders();
+  std::vector<uint8_t>& GetBodyBuffer();
+
+  // adding getters for version and stream body. Clang will complain on Mac if we have unused fields
+  // in a class
+  uint16_t GetMayorVersion() { return m_mayorVersion; }
+  uint16_t GetMinorVersion() { return m_minorVersion; }
+  http::BodyStream* GetBodyStream() { return m_bodyStream; }
 };
 
 class Client
 {
 public:
-  static Response send(Request& request);
+  static Response* Send(Request& request);
```

Comment:
@BillyONeal : The response travels up the pipeline and so it is returned, and returned, and returned. If it's by-value, then it will be copied many times. So, I think it should live on the heap and a pointer or reference to it should be returned. Or, a unique_ptr or something like that.

Reply:
@JeffreyRichter It seems like all such cases should be eligible for the Named Return Value Optimization meaning there is no copy. In the worst case there is a move but that's probably inexpensive.

Reply:
https://gcc.godbolt.org/z/5LxgKN Example, no copies.

Reply:
OK, looks good to me then

Reply:
Sadly it looks like msvc needs the optimizer for this one :(. It's still only a move construction though; we should endeavor to make that relatively inexpensive:

```
C:\Users\billy\Desktop>type demo.cpp
#include <stdio.h>

struct NoisyResponse {
    NoisyResponse() { puts("default constructed"); }
    NoisyResponse(const NoisyResponse&) { puts("copy constructed"); }
    NoisyResponse(NoisyResponse&&) { puts("move constructed"); }
    ~NoisyResponse() { puts("destroyed"); }
    NoisyResponse operator=(const NoisyResponse&) = delete;
    NoisyResponse operator=(NoisyResponse&&) = delete;
};

struct IPipelineStage {
    virtual NoisyResponse Process() const = 0;
};


struct Stage1 : IPipelineStage {
    NoisyResponse Process() const {  NoisyResponse instance; return instance; }
};

struct Stage2 : IPipelineStage {
    NoisyResponse Process() const override {
        Stage1 next{};
        return next.Process();
    }
};

struct Stage3 : IPipelineStage {
    NoisyResponse Process() const override {
        Stage2 next{};
        return next.Process();
    }
};

int main() {
    Stage3 pipeline;
    (void)pipeline.Process();
}

C:\Users\billy\Desktop>cl /EHsc /W4 /WX /nologo demo.cpp
demo.cpp

C:\Users\billy\Desktop>.\demo.exe
default constructed
move constructed
destroyed
destroyed

C:\Users\billy\Desktop>cl /EHsc /W4 /WX /nologo /std:c++latest demo.cpp
demo.cpp

C:\Users\billy\Desktop>.\demo.exe
default constructed
move constructed
destroyed
destroyed

C:\Users\billy\Desktop>cl /EHsc /W4 /WX /nologo /O2 demo.cpp
demo.cpp

C:\Users\billy\Desktop>.\demo.exe
default constructed
destroyed

C:\Users\billy\Desktop>
```

Reply:
@BillyONeal , For implementing this Send() method, I have a transport Class that uses libcurl.
This is for having one instance per request and each instance to maintain its own state.
So Response becomes an internal private field from that class, and it needs to be instantiated when that class is (need to have a default constructor).
However, we don't wan to allow having a default constructor for Response. We want to know that every time a Response is created, we already have a network response to model.

If it is not a private field, Response should need to be created inside the Send() method (stack) to be returned witch makes setting curl headers/body callbacks functions complicated.

So, maybe we can have a quick call and you can advise by looking at code?

Reply:
>However, we don't wan to allow having a default constructor for Response. We want to know that every time a Response is created, we already have a network response to model.

I guess I disagree with this design direction. IMO either the response should be fully formed when it is constructed, or it should not and you should take advantage that it does not, but what you have here is a half measure where some header parts are present upon construction but the response still needs to be edited after the fact.

Reply:
Yes, what you say is true. But, just because we don't create it fully-formed, doesn't mean we should not do as much as we can up front. Once we get things working, we can possibly improve this so we don't fully form the Response unless we have all the headers/body.

Reply:
well, with current design, we know that the minimum response instance that can be created (with the first headers or a network response [status code, version, etc]) is a valid and ready to use Response. 
Yes, later we can add more headers and/or a body but it is still valid.

However, I am not against having a default constructor and documenting that Response can sometimes be in a "building state" where it might not make sense at all.
Or I would also like adding a ResponseBuilder() strategy. But that would require we adopt this programming model into all SDK to use builders for any instance.
```cpp
auto response =  ResponseBuilder()
  .setStatusCode(...)
  .addHeader(...)
  .addHeader(...)
  .build();
```

I think there are PROS and CONS for all scenarios, so, we just need to take a decision for how will we have it